hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/ListStatusRemoteIterator.java

```diff
@@ -0,0 +1,149 @@
+import org.slf4j.LoggerFactory;
+
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.RemoteIterator;
+import org.apache.hadoop.fs.azurebfs.AzureBlobFileSystemStore;
+
+public class ListStatusRemoteIterator implements RemoteIterator<FileStatus> {
+
+  private static final Logger LOG = LoggerFactory
+      .getLogger(ListStatusRemoteIterator.class);
+
+  private static final boolean FETCH_ALL_FALSE = false;
+  private static final int MAX_QUEUE_SIZE = 10;
+
+  private final Path path;
+  private final AzureBlobFileSystemStore abfsStore;
+  private final ArrayBlockingQueue<Iterator<FileStatus>> iteratorsQueue;
+  private final Object asyncOpLock = new Object();
+
+  private boolean firstBatch = true;
+  private boolean isAsyncInProgress = false;
+  private String continuation;
+  private Iterator<FileStatus> currIterator;
+  private IOException ioException;
+
+  public ListStatusRemoteIterator(final Path path,
+      final AzureBlobFileSystemStore abfsStore) {
+    this.path = path;
+    this.abfsStore = abfsStore;
+    iteratorsQueue = new ArrayBlockingQueue<>(MAX_QUEUE_SIZE);
+    currIterator = Collections.emptyIterator();
+    fetchBatchesAsync();
+  }
+
+  @Override
+  public boolean hasNext() throws IOException {
+    if (currIterator.hasNext()) {
+      return true;
+    }
+    updateCurrentIterator();
+    return currIterator.hasNext();
+  }
+
+  @Override
+  public FileStatus next() throws IOException {
+    if (!this.hasNext()) {
+      throw new NoSuchElementException();
+    }
+    return currIterator.next();
+  }
+
+  private void updateCurrentIterator() throws IOException {
+    fetchBatchesAsync();
+    synchronized (this) {
+      if (iteratorsQueue.isEmpty()) {
+        if (ioException != null) {
+          throw ioException;
+        }
+        if (isListingComplete()) {
+          return;
+        }
+      }
+    }
+    try {
+      currIterator = iteratorsQueue.take();
+      if (!currIterator.hasNext() && !isListingComplete()) {
+        updateCurrentIterator();
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      LOG.error("Thread got interrupted: {}", e);
+    }
+  }
+
+  private boolean isListingComplete() {
+    return !firstBatch && (continuation == null || continuation.isEmpty());
+  }
+
+  private void fetchBatchesAsync() {
+    CompletableFuture.runAsync(() -> asyncOp());
+  }
+
+  private void asyncOp() {
+    if (isAsyncInProgress) {
+      return;
+    }
+    synchronized (asyncOpLock) {
+      if (isAsyncInProgress) {
+        return;
+      }
+      isAsyncInProgress = true;
+    }
+    try {
+      while (!isListingComplete() && iteratorsQueue.size() <= MAX_QUEUE_SIZE) {
+        addNextBatchIteratorToQueue();
+      }
+    } catch (IOException e) {
+      ioException = e;
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      LOG.error("Thread got interrupted: {}", e);
```

Comment:
is it right to catch interrupted exception?

Reply:
Yes. Also we need to set the interrut flag again.